File: Code_cpp/98_tree-medium_验证二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=98 lang=cpp
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
// };
class Solution {
public:
    bool helper(TreeNode* root, long long highest, long long lowest){
        if(root == nullptr)
            return true;
        if(root->val<=lowest || root->val>=highest)
            return false;
        
        
        return
            helper(root->right,highest,root->val)&&
            helper(root->left,root->val,lowest);

    }
    bool isValidBST(TreeNode* root) {
        return helper(root, LONG_MAX,LONG_MIN);
    }
};
// @lc code=end

File: Code_java/98_tree-medium_验证二叉搜索树.java
/*
 * @lc app=leetcode.cn id=98 lang=java
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
// public class TreeNode {
//     int val;
//     TreeNode left;
//     TreeNode right;

//     TreeNode() {
//     }

//     TreeNode(int val) {
//         this.val = val;

//     TreeNode(int val, TreeNode left, TreeNode right) {
//         this.val = val;
//         this.left = left;
//         this.right = right;
//     }
// }

class Solution {

    public boolean helper(TreeNode root, long max, long min) {
        if (root == null) {
            return true;
        }
        if(root.val>=max || root.val<=min){
            return false;
        }

        return 
            helper(root.left, Math.min(root.val,max), min) 
            && helper(root.right, max, Math.max(root.val,min));
    }

    public boolean isValidBST(TreeNode root) {
        if (root == null) {
            return true;
        }
        return 
            helper(root.left, root.val, Long.MIN_VALUE) 
            && helper(root.right, Long.MAX_VALUE, root.val);
    }
}
// @lc code=end
